keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,1128 @@
+  resources within the zone if this becomes a concern.
+
+#### Story 3
+* **Requirement:** I prioritize application availability and want to minimize the
+  risk of outages due to localized overload. I'm willing to accept potentially
+  higher costs associated with cross-zone traffic distribution.
+* **Solution:** Set `routingPreference=PreferSpread`
+* **Effect:** The Kubernetes implementation will try to distribute traffic as
+  equally as possible across endpoints, potentially spanning multiple zones or
+  regions. This can improve resilience but might lead to increased network
+  traffic costs.
+
+#### Story 4
+* **Requirement:** As a developer deploying applications across multiple
+  Kubernetes environments, I want a consistent way to express my routing
+  preferences. 
+* **Solution:** Using one of the available standard values in the
+  `routingPreference` field will allow the customer to use the same Helm Chart
+  configurations with greater confidence regardless of the underlying Kubernetes
+  implementation. This simplifies their deployment process and reduces the
+  complexity of managing cross-cluster applications.
+
+#### Story 5
+* **Requirement:** I have some other precise preferences for how traffic should
+  be routed, and I know that my chosen implementation supports the desired
+  preference. 
+* **Solution:** Set `routingPreference=<domain>/<heuristicName>` (where
+  `<domain>` and `<heuristicName>` are provided by your implementation).
+* **Effect:** The Kubernetes implementation will apply the specified routing
+  heuristic. It's important to note that the precise behavior of
+  implementation-specific heuristics might vary.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Standard Heuristic Implementation (kube-proxy dataplane)
+
+kube-proxy based dataplane (along with EndpointSlice controller, within
+kube-controller-manager as the control plane) will support the three standard
+routing preferences (`Default`, `PreferEqualSpread`, `PreferZone`).
+
+#### `Default` and `PreferEqualSpread`
+* Initially, kube-proxy will treat the `Default` preference the same as
+  `PreferEqualSpread`
+* This leverages existing implementation, requiring no major changes.
+
+#### `PreferZone`
+* This preference will be implemented by the use of Hints within EndpointSlices.
```

Comment:
Somewhere in this KEP we probably want to mention how we're going to change the existing kube-proxy logic around handling hints: https://github.com/kubernetes/kubernetes/blob/2363cdcc399cbf428210efb2c51575ddcad2b84a/pkg/proxy/topology.go#L151-L155. 

I tend to think that we should remove any check for an annotation or field on the Service and instead honor hints if _every_ endpoint for that Service has a hint specified.

Reply:
Oh good catch! 
For some reason I was under the impression that we had already made that change but turns out we haven't. I've added a section for this.